### ISSUE TITLE
Allow subset sling assets

### ISSUE THREAD

**Issue by edsoncezar16** [dagster-io/dagster#20215]
### What's the use case?

The current (0.22.8) version of the sling integration does not have a parameter that controls whether we allow subsetting of the resulting AssetsDefinition. Therefore, the only way to materialize a specific stream is to define a replication config with only that stream. That is easy to do but cumbersome to scale, a simple parameter might be a better idea.

### Ideas of implementation

add a paramater "can_subset" to the sling_assets decorator that controls the corresponding paramater in the  internal multi_asset object

![sling-assets](https://github.com/dagster-io/dagster/assets/20863594/e44fa030-ac81-4547-96e8-b00684587574)

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by edsoncezar16** [comment]
I apologize for the previous answer, it DOES NOT solve the issue: all that was done was to change the way dagster interprets what was executed, not what was executed in itself.

Making a parallel with the dbt integration, which does support subsetting, there is an internal mechanism in the library to create a selection argument to pass in a dbt --select flag, for instance:

![Screenshot from 2024-04-08 21-05-15](https://github.com/dagster-io/dagster/assets/20863594/b9c9086e-acf9-4d40-ae33-035952ae7078)

In this way, when we subset the asset selection, we not only change the way dagster sees what is happening, but we also change what is happening so that what dagster sees is what was *de facto* executed. 

Therefore, a possible path to address this issue would be to change the sling cmd that is executed inside replicate based on the execution context, perhaps implementing a homonymous method and using something inside sling replicate:

```python
replication_config = validate_replication(replication_config)
subsetted_replication_config = get_replication_subset_for_context(*args, **kwargs)

...
```

I hope to come soon with new ideas on this topic.

**Comment by Daniel-Vetter-Coverwhale** [comment]
What does this being a multi-asset buy us over just making each stream it's own actual asset? I was doing that before, and when they had the same auto-materialize policy/cron schedule they wound up getting fired off in the same job if that's the main benefit. 

I was going to put this question on the RFC, but figured it might be more relevant to put it here.

**Comment by edsoncezar16** [comment]
@Daniel-Vetter-Coverwhale we are currently doing that with success (I will contribute the main code snippets next), which we call `bare_sling_assets`.

However, a multi-asset approach is much easier to escalate. The caveat is that an asset is essentially an op with nice metadata and amazing visual capabilities. Therefore it does not suffice to change metadata (i.e. AssetSpec parameters), we also have to change how the op changes at runtime based on the selected asset keys in the AssetExecutionContext.

In the dbt integration, this op customization at runtime is already present in the library, whereas for the sling integration, it is not. Therefore, the naive approach "just change a parameter" does not cut it.
Oddly enough, by making the necessary adjustments to implement the `bare_sling_assets` I believe entirely similar steps would do the trick for the multi-asset framework. However, I am still working on it.

**Comment by edsoncezar16** [comment]
This is a decorator that would allow subset, but also pay attention on how the actual compute function that executes at runtime reads from the context to adapt the `replication.yaml` on the fly - this is analogous to the `dbt select` 

```python
def bare_sling_assets(
    replication_config: SlingReplicationParam,
    custom_dagster_sling_translator: CustomDagsterSlingTranslator,
    partitions_def: Optional[PartitionsDefinition] = None,
    backfill_policy: Optional[BackfillPolicy] = None,
    op_tags: Optional[Mapping[str, Any]] = None,
) -> Callable[[Callable[..., Any]], list[AssetsDefinition]]:
    """
    Maps 1:1 Sling streams to Dagster Assets.
    Workaround for currently open issue: https://github.com/dagster-io/dagster/issues/20215
    """
    replication_config = validate_replication(replication_config)

    def inner(fn: Callable[..., Any]) -> list[AssetsDefinition]:
        asset_defs = []
        stream_definitions = get_streams_from_replication(replication_config)
        for stream_definition in stream_definitions:
            name = stream_definition.get("name")
            config = custom_dagster_sling_translator.get_rectified_stream_config(
                stream_definition
            )
            replication_config["streams"] = {name: config}
            asset_key = custom_dagster_sling_translator.get_asset_key(stream_definition)

            asset_def = asset(
                key=asset_key,
                deps=custom_dagster_sling_translator.get_deps_asset_key(
                    stream_definition
                ),
                metadata={
                    **custom_dagster_sling_translator.get_metadata(stream_definition),
                    METADATA_KEY_REPLICATION_CONFIG: replication_config,
                    METADATA_KEY_TRANSLATOR: custom_dagster_sling_translator,
                },
                description=custom_dagster_sling_translator.get_description(
                    stream_definition
                ),
                compute_kind="sling",
                partitions_def=partitions_def,
                op_tags=op_tags,
                group_name=custom_dagster_sling_translator.get_group_name(
                    stream_definition
                ),
                output_required=True,
                freshness_policy=custom_dagster_sling_translator.get_freshness_policy(
                    stream_definition
                ),
                auto_materialize_policy=custom_dagster_sling_translator.get_auto_materialize_policy(
                    stream_definition
                ),
                backfill_policy=backfill_policy,
                code_version=non_secure_md5_hash_str(str(replication_config).encode()),
            )(fn)

            asset_defs.append(asset_def)
        return asset_defs

    return inner

```

This compute is specific to an S3 source that will be triggered via sensors, but the same principle applies to the general case: the main idea is to take advantage of asset metadata to construct a `replication_config` adequate to the current exec context.
```python
def compute(
    context: AssetExecutionContext,
    sling: SlingResource,
    config: SlingS3StreamsConfig,
) -> Iterator[AssetMaterialization | MaterializeResult]:
    asset_key = context.asset_key
    assets_def = context.assets_def
    asset_metadata = assets_def.metadata_by_key[asset_key]
    stream_config = asset_metadata["stream_config"].data
    run_config = context.run_config
    if run_config:  # triggered via sensor
        asset_op_config = run_config["ops"][asset_key.to_user_string()]["config"]
        new_files = asset_op_config["new_files"]
        new_streams = {}
        for file in new_files:
            new_streams[file] = stream_config
            replication_config["streams"] = new_streams
    else:
        s3_pattern = asset_metadata["s3_pattern"].value
        replication_config["streams"] = {s3_pattern: stream_config}
    yield from sling.replicate(
        context=context,
        replication_config=replication_config,
        dagster_sling_translator=custom_dagster_sling_translator,
        debug=True,
    )
```

**Comment by edsoncezar16** [comment]
Hi folks, I finally managed to get a clean solution for this issue, thoroughly validated by materializing assets in the UI and via individual asset sensors as well:

The core idea, as pointed out in an earlier comment, is to use the `subsettable_sling_assets` decorator in a compute function that modifies the replication at runtime based on the exec context. Here it follows the current solution:

```python
@subsettable_sling_assets(
    replication_config=replication_path,
    dagster_sling_translator=custom_dagster_sling_translator
)
def compute(
    context: AssetExecutionContext,
    sling: SlingResource,
    config: SlingS3StreamsConfig,
) -> Iterator[AssetMaterialization | MaterializeResult]:
    replication_config = validate_replication(replication_path)
    replication_config["streams"] = get_replication_streams_for_context(context)
    yield from sling.replicate(
        context=context,
        replication_config=replication_config,
        dagster_sling_translator=custom_dagster_sling_translator,
        debug=True,
    )
```

The ancilla function is this one:

```python
def get_replication_streams_for_context(context: AssetExecutionContext):
    """
    Computes the sling replication streams config for a given execution context
    of a subsettable sling assets definition.
    """
    context_streams = {}
    assets_def = context.assets_def
    run_config = context.run_config
    if run_config:  # triggered via sensor
        assets_op_config = run_config["ops"][assets_def.op.name]["config"]
        asset_key = AssetKey(assets_op_config["asset_key_str"])
        new_files = assets_op_config["new_files"]
        asset_metadata = assets_def.metadata_by_key[asset_key]
        stream_config = asset_metadata["stream_config"].data
        for file in new_files:
            context_streams[file] = stream_config
    else:
        for asset_key in context.selected_asset_keys:
            asset_metadata = assets_def.metadata_by_key[asset_key]
            stream_config = asset_metadata["stream_config"].data
            s3_pattern = asset_metadata["s3_pattern"].value
            context_streams[s3_pattern] = stream_config
    return context_streams
```


At this point, it has some specificities for how we are using s3 sensors, but it can easily be extended for a generic case by transferring the logic to the sensor evaluation function and providing the `context_streams` directly in the run_config:
```python
def get_replication_streams_for_context(context: AssetExecutionContext):
    context_streams = {}
    assets_def = context.assets_def
    run_config = context.run_config
    if run_config:  # triggered via sensor
        assets_op_config = run_config["ops"][assets_def.op.name]["config"]
        context_streams = assets_op_config["context_streams"]
       ...
    return context_streams
```

@PedramNavid I would be more than willing to contribute a PR to fix this issue, as well as  issue #21075 . Please let me know if that would be welcome on your part.

**Comment by PedramNavid** [comment]
Hi @edsoncezar16, would definitely welcome a PR ! If you're in our community Slack, message me there and I'll add you to our OSS Contributors channel as well.

**Comment by edsoncezar16** [comment]
Fixed by PR #21390 .
